src/structurizr/model/container.py

```diff
@@ -70,7 +71,7 @@ class Container(StaticStructureElement):
         technology: The technology associated with this container
                     (e.g. Apache Tomcat).
         tags: A comma separated list of tags associated with this container.
-        components: The set of components within this container.
+        components: The set of components within this container.  Do not modify directly.
```

Comment:
How concerned are we about people modifying components directly?  We could provide a read-only property (returning a read-only set or iterable) instead, but is it worth it?

Reply:
As much as it's possible in Python I would like to disallow modifying components (and other container attributes) directly indeed. It is extra effort yes but it has the well known benefit that we can modify the underlying container type (list, set, dict) to our liking whenever we please.

Reply:
Agreed.  This will be quite a big change across many classes if we want to enforce it consistently (especially as it's not just about making the property read-only, but also that to the outside world it looks like a read-only collection).

Another example is in BigBank.py you can see adding to the tags collection directly.

Reply:
I've just refactored this specific case (Container.components) to try out one approach.  See what you think.